### PR TITLE
Default CONJUR_VERSION behavior refactored and tested

### PR DESCRIPTION
**Pull request by BradleyBoutcher**
When building a configuration from the environment, we now verify
that the input value for the Conjur version is '4' or '5'. If empty,
it will use a default value of '5'. If any other value is provided, we
throw an error.

- Performed a minor refactoring of NewFromEnv to allow functional options as arguments
- Created unit tests for conjur versions, with a framework which allows for additional unit tests for any of the features in config.go

### CODE REVIEW

File: pkg/authenticator/config/config.go [hunk 1]
```diff
@@ -28,12 +29,37 @@ const (
...
+}
+
+// DefaultNewFromEnv returns NewFromEnv with the default functional options
+func DefaultNewFromEnv() (*Config, error) {
+	return NewFromEnv(DefaultConfigOptions)
+}
+
 // New returns a new authenticator configuration object
-func NewFromEnv() (*Config, error) {
+func NewFromEnv(opts Options) (*Config, error) {
```
Comment (sgnn7) [inline pkg/authenticator/config/config.go]: This is a breaking change in almost all packages that use our authenticator. I would highly avoid changing this method's signature. As far as I can tell, this was changed to accommodate the tests but this can be tested without this change I think.

File: pkg/authenticator/config/config.go [hunk 2]
```diff
@@ -28,12 +29,37 @@ const (
...
+}
+
+// DefaultConfigOptions implement the methods defined within config.go
+var DefaultConfigOptions = Options{
+	readSSLCert,
+	NewUsername,
+}
+
+// DefaultNewFromEnv returns NewFromEnv with the default functional options
+func DefaultNewFromEnv() (*Config, error) {
```
Comment (sgnn7) [inline pkg/authenticator/config/config.go]: Normally, the prefix for Golang constructors is `New` so if a change is needed, it should contain methods similar to `NewFromEnv` and `NewFromEnvWithOptions`.

File: pkg/authenticator/config/config.go [hunk 3]
```diff
@@ -28,12 +29,37 @@ const (
...
+type Options struct {
+	ReadSSLCert ReadSSLCertFunc
+	NewUserName NewUserNameFunc
+}
+
+// DefaultConfigOptions implement the methods defined within config.go
+var DefaultConfigOptions = Options{
+	readSSLCert,
+	NewUsername,
+}
```
Comment (sgnn7) [inline pkg/authenticator/config/config.go]: I could be wrong but I would double-check that you really need any of these types to blackbox-test the authenticator - I think they might not be needed.

File: pkg/authenticator/config/config.go [hunk 4]
```diff
@@ -68,9 +94,12 @@ func NewFromEnv() (*Config, error) {
...
-	conjurVersion := os.Getenv("CONJUR_VERSION")
-	if len(conjurVersion) == 0 {
-		conjurVersion = "5"
+	conjurVersion := DefaultConjurVersion
+	// Only '4' and '5' are supported, with empty values using the default of '5'
+	if version := os.Getenv("CONJUR_VERSION"); version == "4" {
+		conjurVersion = version
+	} else if version != "5" && version != "" {
+		return nil, log.RecordedError(log.CAKC021E, fmt.Errorf("invalid conjur version"))
 	}
```
Comment (sgnn7) [inline pkg/authenticator/config/config.go]: This feels like it should be a `select` rather than an `if`

File: pkg/authenticator/config/config_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/cyberark/conjur-authn-k8s-client/pkg/log"
+	"os"
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+var mockSuccessfulOptions = Options{
+	successfulMockReadSSLCertFunc,
+	 successfulMockNewUserNameFunc,
```
Comment (sgnn7) [inline pkg/authenticator/config/config_test.go]: nit: Indentation is inconsistent

File: pkg/authenticator/config/config_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+var environmentValues = map[string]string{
+	"CONJUR_SSL_CERTIFICATE": "testSSLCert",
+	"CONJUR_CERT_FILE": "testSSLCertFile.txt",
+	"CONJUR_AUTHN_URL": "filepath",
+	"CONJUR_ACCOUNT": "testAccount",
+	"CONJUR_AUTHN_LOGIN": "host",
+	"MY_POD_NAMESPACE": "testNameSpace",
+	"MY_POD_NAME": "testPodName",
+	"CONJUR_VERSION": "",
```
Comment (sgnn7) [inline pkg/authenticator/config/config_test.go]: nit: Would be good to sort these since it's hard to find a variable in an unordered list

File: pkg/authenticator/config/config_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func TestAuthenticator(t *testing.T) {
+	// Set default environment variables
+	for key, value := range environmentValues {
+		err := os.Setenv(key, value)
+		if err != nil {
+			panic(err)
+		}
+	}
```
Comment (sgnn7) [inline pkg/authenticator/config/config_test.go]: This should have a `defer`ed unsetting of those variables to not pollute other tests

File: pkg/authenticator/config/config.go [hunk 8]
```diff
@@ -68,9 +94,12 @@ func NewFromEnv() (*Config, error) {
 
 	containerMode := os.Getenv("CONTAINER_MODE")
 
-	conjurVersion := os.Getenv("CONJUR_VERSION")
-	if len(conjurVersion) == 0 {
-		conjurVersion = "5"
+	conjurVersion := DefaultConjurVersion
+	// Only '4' and '5' are supported, with empty values using the default of '5'
+	if version := os.Getenv("CONJUR_VERSION"); version == "4" {
+		conjurVersion = version
```
Comment (izgeri) [inline pkg/authenticator/config/config.go]: I understand that this gives the intended result (the tests make that clear, if nothing else)

but the logic seems odd to me, because it is not clear that when CONJUR_VERSION is set to 5 we also set version = CONJUR_VERSION

I know it is simpler the way you have it written, and achieves the same end result - but I feel it could be clearer what the intended behavior is, even if it ends up being a bit more verbose.

but not a dealbreaker, in any case

File: pkg/authenticator/config/config.go (hunk 3 above)
Comment (BradleyBoutcher) [reply pkg/authenticator/config/config.go]: The issue lies mostly with readSSLCert requiring an actual file to read, and NewUsername is also an external method which requires a specific value, which is difficult to mock out well. If possible I felt this functional option implementation would make it easier to test failure paths in NewFromEnv and to add tests for the other code paths while maintaining readability. Of course, I very much don't want to break existing functionality, but I'll need help figuring out the best way to do this.

Comment (BradleyBoutcher) [reply pkg/authenticator/config/config.go]: For now I just swapped the signatures in a quick fix, but I definitely want to hear more about better ways we could do this

File: pkg/authenticator/config/config.go (hunk 8 above)
Comment (diverdane) [reply pkg/authenticator/config/config.go]: I agree, the logic would be clearer and more extensible to future versions with a switch, something like this (leave out the comments if it's too verbose):
```
	conjurVersion := os.Getenv("CONJUR_VERSION")
	switch conjurVersion {
	case "4": // Valid
	case "5": // Valid
	case "":  // Use default
		conjurVersion = DefaultConjurVersion
	default: // Unknown version
		return nil, log.RecordedError(log.CAKC021E, fmt.Errorf("invalid conjur version"))
	}
```

File: pkg/authenticator/config/config_test.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+	Convey("NewFromEnv", t, func() {
```
Comment (diverdane) [inline pkg/authenticator/config/config_test.go]: This is a matter of personal preference, but table driven test could be used here to consolidate the corner cases that are considered, e.g.:
```
func TestAuthenticator(t *testing.T) {
	// Set default environment variables
	for key, value := range environmentValues {
		err := os.Setenv(key, value)
		if err != nil {
			panic(err)
		}
	}

	TestCases := []struct {
		description string
		envVersion  string
		expVersion  string
		expErrStr   string
	}{
		{
			description: "Succeeds if version is 4",
			envVersion:  "4",
			expVersion:  "4",
			expErrStr:   "",
		},
		{
			description: "Succeeds if version is 5",
			envVersion:  "5",
			expVersion:  "5",
			expErrStr:   "",
		},
		{
			description: "Sets the default version for an empty value",
			envVersion:  "",
			expVersion:  DefaultConjurVersion,
			expErrStr:   "",
		},
		{
			description: "Returns error if version is invalid",
			envVersion:  "3",
			expVersion:  "",
			expErrStr:   fmt.Sprintf(log.CAKC021E, "invalid conjur version"),
		},
	}

	Convey("NewFromEnv", t, func() {
		for _, tc := range TestCases {
			Convey(tc.description, func() {
				_ = os.Setenv("CONJUR_VERSION", tc.envVersion)

				config, err := NewNewFromEnv(mockSuccessfulOptions)

				if tc.expErrStr == "" {
					So(err, ShouldBeNil)
					So(config.ConjurVersion, ShouldEqual, tc.expVersion)
				} else {
					So(err, ShouldNotBeNil)
					So(err.Error(), ShouldEqual, tc.expErrStr)
				}
			})
		}
	})
}
```

File: pkg/authenticator/config/config.go [hunk 10]
```diff
@@ -68,9 +99,17 @@ func NewFromEnv() (*Config, error) {
...
 	containerMode := os.Getenv("CONTAINER_MODE")
 
-	conjurVersion := os.Getenv("CONJUR_VERSION")
-	if len(conjurVersion) == 0 {
-		conjurVersion = "5"
+	// Only versions '4' & '5' are allowed, with '5' being used as the default
+	conjurVersion := DefaultConjurVersion
+	switch os.Getenv("CONJUR_VERSION") {
+	case "4":
+		conjurVersion = os.Getenv("CONJUR_VERSION")
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: A slight variation, maybe... Instead of calling os.Getenv() twice, you already know what it's going to be:
```
	case "4":
		conjurVersion = "4"
```

File: pkg/authenticator/config/config.go [hunk 11]
```diff
@@ -28,37 +29,67 @@ const (
...
+	NewUserName NewUserNameFunc
+}
+
+// DefaultConfigOptions implement the methods defined within config.go
+var DefaultConfigOptions = Options{
+	readSSLCert,
+	NewUsername,
+}
+
+// NewFronEnv returns a config FromEnv with the default functional options
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: Typo, "NewFronEnv"

File: pkg/authenticator/config/config.go [hunk 12]
```diff
@@ -28,37 +29,67 @@ const (
...
+	caCert, err := opts.ReadSSLCert()
 	if err != nil {
 		return nil, log.RecordedError(log.CAKC021E, err.Error())
 	}
 
 	// Load configuration from the environment
 	authnURL := os.Getenv("CONJUR_AUTHN_URL")
 	account := os.Getenv("CONJUR_ACCOUNT")
-	authnLogin, err := NewUsername(os.Getenv("CONJUR_AUTHN_LOGIN"))
+	authnLogin, err := opts.NewUserName(os.Getenv("CONJUR_AUTHN_LOGIN"))
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: I believe that the `NewUserName` call here doesn't need to be mocked for UT. To convince myself, I put back the original call here on line 92, and the UT ran fine.

File: pkg/authenticator/config/config.go [hunk 13]
```diff
@@ -28,88 +29,109 @@ const (
...
+	"CONJUR_AUTHN_LOGIN",
+	"MY_POD_NAMESPACE",
+	"MY_POD_NAME",
+}
+
+// NewFromEnvFunc defines the method for creating a new config from the env
+type NewFromEnvFunc func(Options) (*Config, error)
+
+// ReadSSLCertFunc defines the method for reading an SSL Certificate from the env
+type ReadSSLCertFunc func() ([]byte, error)
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: Instead of mocking the `readSSLCert` function, you could get better coverage by mocking the lower level file reader inside the `readSSLCert` function itself, i.e. this line:
```
return ioutil.ReadFile(SSLCertPath)
```
The way to do this would be to define a `fileReader` interface, and add a `fileReader` interface as an argument to the `FromEnv()` function signature (no need to use functional options since there's only one option, and it's a simple interface). Within `FromEnv()`, you would pass on that interface to the call to `readSSLCert()`, and inside `readSSLCert()`, you would do the reading of the file using the interface passed in.
Then in your UT, you would pass in a buffer or some other IO reader as a mock.

File: pkg/authenticator/config/config.go [hunk 14]
```diff
@@ -28,88 +29,109 @@ const (
...
+	config.SSLCertificate, err = opts.ReadSSLCert()
 	if err != nil {
 		return nil, log.RecordedError(log.CAKC021E, err.Error())
 	}
 
 	// Load configuration from the environment
-	authnURL := os.Getenv("CONJUR_AUTHN_URL")
-	account := os.Getenv("CONJUR_ACCOUNT")
-	authnLogin, err := NewUsername(os.Getenv("CONJUR_AUTHN_LOGIN"))
+	config.Username, err = opts.NewUserName(os.Getenv("CONJUR_AUTHN_LOGIN"))
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: Apologies if this comment shows up twice, but I didn't see it showing up in the file diffs:
I believe that the NewUserName call here doesn't need to be mocked for UT. To convince myself, I put back the original call here on line 92, and the UT ran fine.

File: pkg/authenticator/config/config_test.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+					So(err.Error(), ShouldEqual, tc.expErrStr)
+				}
+			})
+		}
+	})
+}
+
+func successfulMockReadFile(filename string) ([]byte, error) {
+	return []byte{}, nil
+}
```
Comment (diverdane) [inline pkg/authenticator/config/config_test.go]: Github complains: Missing newline on line 95.

File: pkg/authenticator/config/config.go [hunk 16]
```diff
@@ -28,91 +29,94 @@ const (
...
+
+// NewFromEnv returns a config FromEnv with the default functional options
 func NewFromEnv() (*Config, error) {
-	var err error
+	return FromEnv(ioutil.ReadFile)
+}
+
+// FromEnv returns a new authenticator configuration object
+func FromEnv(readFileFunc ReadFileFunc) (*Config, error) {
+	var err error = nil
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: Can drop the nil, since that's the zero value for an error type, i.e. just use `var err error`.

File: pkg/authenticator/config/config.go [hunk 17]
```diff
@@ -28,91 +29,94 @@ const (
 	DefaultClientCertPath = "/etc/conjur/ssl/client.pem"
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: Some linters are going to complain about `DefaultClientCertPath` not having a comment for an exported variable. Same with the other Default* consts below. You could satisfy linter for all of these with a comment before the `const (`, e.g. :
```
// Default settings (this comment added to satisfy linter)
const (
```

File: pkg/authenticator/config/config.go [hunk 18]
```diff
@@ -28,91 +29,94 @@ const (
...
+	"CONJUR_ACCOUNT",
+	"CONJUR_AUTHN_LOGIN",
+	"MY_POD_NAMESPACE",
+	"MY_POD_NAME",
+}
+
+// ReadFileFunc defines the interface for reading an SSL Certificate from the env
+type ReadFileFunc func(filename string) ([]byte, error)
+
+// NewFromEnv returns a config FromEnv with the default functional options
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: "with the default functional options" should be "using the standard file reader" or something like that.

File: pkg/authenticator/config/config.go [hunk 19]
```diff
@@ -25,102 +26,124 @@ type Config struct {
 }
 
 const (
+	// DefaultClientCertPath (this comment added to satisfy linter)
```
Comment (diverdane) [inline pkg/authenticator/config/config.go]: You could have used one comment at the head of this block (before the `const (` line) to handle all of the exported consts, but this works.

**Comment by BradleyBoutcher** [comment]
@sgnn7 
Done
